Add README (#5)
university-of-york/uoy-config-funnelback-courses: README.md

# UOY-CONFIG-FUNNELBACK-COURSES
This repo contains configuration for the University of York's course search powered by [Funnelback](https://www.funnelback.com/).

## Funnelback

### Search Template
Search templates are *.ftl files that determine the format of results from Funnelback.

### Collection
A Funnelback collection is a set of data that can be searched. It is possible for collections to contain other collections, 

> as the york-uni-web main York collection contains others.

Reviewer: york-uni-web is actually a plain web collection, whereas the Campus map collection is a 'meta' collection containing others

```suggestion
as the york-uni-campusmap collection contains others.
```

The course collection is called *york-uni-courses*.

### Development
Useful resources:
* [Funnelback Documentation](https://docs.funnelback.com/)
* [OpenAPI Documentation](https://swagger.io/specification/)

### Testing
The York-uni-courses collection has two profiles, for 'Live' and 'Preview'. The preview profile can be used for testing 
purposes. For details on how to deploy files to the preview profile, see 'Deployment'.

### Deployment
Deployment of configuration files to Funnelback cannot be done programmatically, and must be done using the web interface.

While logged in to the Funnelback admin console, find the *york-uni-courses* collection and select 'Browse Collection Configuration Files'.
From here you will see a list of configuration files for each profile. In order to publish a configuration file to Live, 
it must first be pushed to 'Preview'.

> ####To deploy a file:

Reviewer 2: Extra #

> ####To deploy a file:

Author: ```suggestion
#### To deploy a file:
```

> Click on the name of the file you want to deploy in the preview section. This takes you to a visual editor. You will need
> to copy the code from the matching file in Source Control to overwrite the file, and then click **Save**.
> 
> Test that the file is working by running an example search, then when you are satisfied return to the config list and click **Publish** to publish that file to Live.
> 
> If you are uploading a new file, use the file picker below the 'Preview' profile.
> 

Reviewer: Re-written part of this section to talk about uploading & archiving files
```suggestion
The simplest method of uploading/updating a file is to use the file picker at the bottom of the section. 
There is also a drop-down menu that allows for the creation of new files that indicates the possible filenames that will be recognised by Funnelback.

When uploading a new version of an existing file, the old one will be archived with a date-time stamp within Funnelback.
This allows for a change to be quickly reverted by using the 'Restore' option on an archived version.

Funnelback also provides a visual interface for viewing/editing existing files. 
Click on the name of the file you want to edit in the preview section to open the editor. 
You will need to copy the code from the matching file in Source Control to overwrite the file, and then click **Save**.

Test that the file is working by running an example search, then when you are satisfied return to the config list and click **Publish** to publish that file to Live.
```

From here you will see a list of configuration files for each profile. In order to publish a configuration file to Live, 
it must first be pushed to 'Preview'.

#### To deploy a file:
The simplest method of uploading/updating a file is to use the file picker at the bottom of the section. 
There is also a drop-down menu that allows for the creation of new files that indicates the possible filenames that will be recognised by Funnelback.

When uploading a new version of an existing file, the old one will be archived with a date-time stamp within Funnelback.
This allows for a change to be quickly reverted by using the 'Restore' option on an archived version.

Funnelback also provides a visual interface for viewing/editing existing files. 
Click on the name of the file you want to edit in the preview section to open the editor. 
You will need to copy the code from the matching file in Source Control to overwrite the file, and then click **Save**.

Test that the file is working by running an example search, then when you are satisfied return to the config list and click **Publish** to publish that file to Live.
## OpenAPI Specification

> The OpenAPI specification can be viewed at https://university-of-york.github.io/uoy-config-funnelback-courses. It is served 

Reviewer 3: I get a 404 error on this URL

Author: That will be the case until the gh-pages branch is set up, the URL should be correct though

Reviewer 2: That should be fixed now Nicola.

# UOY-CONFIG-FUNNELBACK-COURSES

> This repo contains configuration for the University of York's course search powered by [Funnelback](https://www.funnelback.com/).

Reviewer 4: Could we expand on this a little? e.g. `This configures a collection of course information within Funnelback, and a JSON API to search this collection. This can be seen at <sample search url>.`

Author: ```suggestion
This repo contains configuration for the University of York's course search powered by [Funnelback](https://www.funnelback.com/). This configures a collection of course information within funnelback, which can be queried as an API returning JSON. This can be seen [here](https://york.funnelback.co.uk/s/search.json?collection=york-uni-courses&query=History&profile=_default_preview&form=course-search)
```

## Funnelback

### Search Template
Search templates are *.ftl files that determine the format of results from Funnelback.

### Collection
A Funnelback collection is a set of data that can be searched. It is possible for collections to contain other collections, 

> as the york-uni-campusmap collection contains others.

Reviewer 4: Going to suggest we be consistent with formatting of collection names across this document

```suggestion
as the `york-uni-campusmap` collection contains others.
```

> The course collection is called *york-uni-courses*.

Reviewer 4: ```suggestion
The course collection is called `york-uni-courses`.
```

### Development
Useful resources:
* [Funnelback Documentation](https://docs.funnelback.com/)

> * [OpenAPI Documentation](https://swagger.io/specification/)

Author: ```suggestion
* [OpenAPI Documentation](https://swagger.io/docs/specification/)
```
Easier to read and more clear

### Testing

> The York-uni-courses collection has two profiles, for 'Live' and 'Preview'. The preview profile can be used for testing 

Reviewer 4: ```suggestion
The `york-uni-courses` collection has two profiles, for 'Live' and 'Preview'. The preview profile can be used for testing 
```

> The York-uni-courses collection has two profiles, for 'Live' and 'Preview'. The preview profile can be used for testing 

Reviewer 4: Can we link to searches using the live and preview templates?

Author: Should we include multiple links to the search or emphasise use of Swagger-UI (whose URL will never change) to evaluate the search?

purposes. For details on how to deploy files to the preview profile, see 'Deployment'.

### Deployment

> Deployment of configuration files to Funnelback cannot be done programmatically, and must be done using the web interface.

Reviewer 4: Suggest we have a link to the Funnelback web interface

Author: ```suggestion
Deployment of configuration files to Funnelback cannot be done programmatically, and must be done using the [web interface](https://york.funnelback.co.uk/search/admin).
```

While logged in to the Funnelback admin console, find the *york-uni-courses* collection and select 'Browse Collection Configuration Files'.
From here you will see a list of configuration files for each profile. In order to publish a configuration file to Live, 
it must first be pushed to 'Preview'.

#### To deploy a file:
The simplest method of uploading/updating a file is to use the file picker at the bottom of the section. 

> There is also a drop-down menu that allows for the creation of new files that indicates the possible filenames that will be recognised by Funnelback.

Reviewer 4: A screenshot is worth a thousand words 🙂 I could see it being really helpful e.g. if I needed to do an emergency deployment and hadn't really used Funnelback before

Author: ```suggestion
There is also a drop-down menu that allows for the creation of new files that indicates the possible filenames that will be recognised by Funnelback.
![File picker](https://user-images.githubusercontent.com/68321515/91434581-f0095480-e85c-11ea-9480-f95f217d6a46.png)
```

When uploading a new version of an existing file, the old one will be archived with a date-time stamp within Funnelback.
This allows for a change to be quickly reverted by using the 'Restore' option on an archived version.

Funnelback also provides a visual interface for viewing/editing existing files. 
Click on the name of the file you want to edit in the preview section to open the editor. 
You will need to copy the code from the matching file in Source Control to overwrite the file, and then click **Save**.

Test that the file is working by running an example search, then when you are satisfied return to the config list and click **Publish** to publish that file to Live.

> ## OpenAPI Specification

Reviewer 4: I think this section should be further up, possibly above the Funnelback section. This is the interface that we're publishing, the output of this repo, and the Funnelback information is everything that goes behind it, so it makes sense that this would be first to me.

If we also mention the API and sample URL in the introduction then I think it ties nicely together at that point - that's just my feeling trying to think of someone coming to this for the first time!

> ## OpenAPI Specification

Reviewer 4: Small suggestion, in case people don't know what OpenAPI specification is/means - we can explain that in the text.

```suggestion
## API Specification
```

> The OpenAPI specification can be viewed at https://university-of-york.github.io/uoy-config-funnelback-courses. It is served 
> here as a static site using a distribution of [Swagger-UI](https://github.com/swagger-api/swagger-ui/tree/master/dist).

Reviewer 4: Better to have descriptive links (as you've done elsewhere!)

```suggestion
The [latest version of the API specification](https://university-of-york.github.io/uoy-config-funnelback-courses) is published and kept up-to-date automatically. 

It is defined using the [OpenAPI specification](https://swagger.io/docs/specification/about/) and is served as a static site using a distribution of [Swagger-UI](https://github.com/swagger-api/swagger-ui/tree/master/dist).
```

> ## The Swagger UI

Reviewer 4: This is all really useful info, I wonder if it might be more suited to the main wiki? It's almost metadata about the repo i.e. how we set up the documentation, and how others could do it in future in their repos, rather than directly related to Funnelback config. Don't know if anyone else has thoughts on this?

### Overview

> The swagger UI allows the openAPI demonstration rendering of a Funnelback search return to be viewed in a way that users will find helpful.

Reviewer 4: ```suggestion
The Swagger UI allows the OpenAPI demonstration rendering of a Funnelback search return to be viewed in a way that users will find helpful.
```

    3)  React Swagger UI:
        This is Swagger UI packaged as a React component for use in React applications.
Current and further information on the Swagger UI and its branches can be found on the Swagger UI repo Readme.md.

### openAPI output
The openAPI location is: /spec/openAPI.yml

### Modifications to the dist folder (swagger-ui folder on our git-hub repository) 
The following steps were undertaken to produce the version we use of the Swagger UI:

1)  The Swagger UI was downloaded from github and a local copy was created. 
2)  The dist folder was renamed to swagger-ui for use in our github repository. 
3)  The index.html that was in the swagger-ui  folder (formally the dist folder) was then moved down a level to the root of the Swagger UI package.
4)  The index.html had the following applied: 
    a)  A change all was undertaken against all paths that had “./” to now read “swagger-ui/”
    b)  The link to the openAPI was changed from the default link to that of the spec folders openAPI.yml 

### Output 

> The team have chosen to use github pages as the way to host the Swagger UI, the output is available at: https://university-of-york.github.io/uoy-config-funnelback-courses/  This is a publicly accessible location. 

Reviewer 4: Suggest we add a Contact section at the bottom - using email addresses as this is a public repo.

```suggestion
The team have chosen to use github pages as the way to host the Swagger UI, the output is available at: https://university-of-york.github.io/uoy-config-funnelback-courses/  This is a publicly accessible location. 

## Contact

- [Digital Platforms and Developments Team](mailto:marketing-support@york.ac.uk)
- [Enterprise Systems Teaching and Learning Team](mailto:esg-teaching-and-learning-group@york.ac.uk)
```

# UOY-CONFIG-FUNNELBACK-COURSES

> This repo contains configuration for the University of York's course search powered by [Funnelback](https://www.funnelback.com/). This configures a collection of course information within funnelback, which can be queried as an API returning JSON. This can be seen [here](https://york.funnelback.co.uk/s/search.json?collection=york-uni-courses&query=History&profile=_default_preview&form=course-search)

Reviewer 4: ```suggestion
This repo contains configuration for the University of York's course search powered by [Funnelback](https://www.funnelback.com/). This configures a collection of course information within Funnelback, which can be queried as an API returning JSON. This can be seen [here](https://york.funnelback.co.uk/s/search.json?collection=york-uni-courses&query=History&profile=_default_preview&form=course-search)
```
